cli/src/cmd/forge/coverage.rs

```diff
@@ -0,0 +1,283 @@
+            let mut project = config.ephemeral_no_artifacts_project()?;
+
+            // Disable the optimizer for more accurate source maps
+            project.solc_config.settings.optimizer.disable();
+
+            project
+        };
+
+        let output = ProjectCompiler::default()
+            .compile(&project)?
+            .with_stripped_file_prefixes(project.root());
+
+        Ok((project, output))
+    }
+
+    /// Builds the coverage map.
+    fn prepare(&self, output: ProjectCompileOutput) -> eyre::Result<(CoverageMap, SourceMaps)> {
+        // Get sources and source maps
+        let (artifacts, sources) = output.into_artifacts_with_sources();
+
+        let source_maps: SourceMaps = artifacts
+            .into_iter()
+            .map(|(id, artifact)| (id, CompactContractBytecode::from(artifact)))
+            .filter_map(|(id, artifact): (ArtifactId, CompactContractBytecode)| {
+                Some((
+                    id,
+                    (
+                        artifact.get_source_map()?.ok()?,
+                        artifact
+                            .get_deployed_bytecode()
+                            .as_ref()?
+                            .bytecode
+                            .as_ref()?
+                            .source_map()?
+                            .ok()?,
+                    ),
+                ))
+            })
+            .collect();
+
+        let mut map = CoverageMap::default();
+        for (path, versioned_sources) in sources.0.into_iter() {
+            // TODO: Make these checks robust
+            // NOTE: We should actually filter out test contracts in the AST
+            // instead of on a source file level. Repositories like Solmate
+            // have a lot of abstract contracts that are being tested, and these
+            // are usually defined in the test files themselves.
+            let is_test = path.ends_with(".t.sol");
+            let is_dependency = path.starts_with("lib");
+            if is_test || is_dependency {
+                continue
+            }
+
+            for mut versioned_source in versioned_sources {
+                let source = &mut versioned_source.source_file;
+                if let Some(ast) = source.ast.take() {
+                    let source_maps: HashMap<String, SourceMap> = source_maps
+                        .iter()
+                        .filter(|(id, _)| {
+                            id.version == versioned_source.version &&
+                                id.source == PathBuf::from(&path)
+                        })
+                        .map(|(id, (_, source_map))| {
+                            // TODO: Deploy source map too?
+                            (id.name.clone(), source_map.clone())
+                        })
+                        .collect();
+
+                    let items = Visitor::new(
+                        source.id,
+                        std::fs::read_to_string(PathBuf::from(&path))?,
```

Comment:
let's wrap all io errors with their paths, because the path is not included in `io::Error`